testutil/containschecker.go

```diff
@@ -150,3 +150,73 @@ func (c *deepContainsChecker) Check(params []interface{}, names []string) (resul
 		return false, fmt.Sprintf("%T is not a supported container", container)
 	}
 }
+
+type deepUnsortedMatchChecker struct {
+	*check.CheckerInfo
+}
+
+// DeepUnsortedMatches checks if two containers contain the same elements in
+// the same number (but possibly different order) using DeepEqual. The container
+// can be an array, a slice or a map.
+var DeepUnsortedMatches check.Checker = &deepUnsortedMatchChecker{
+	&check.CheckerInfo{Name: "DeepUnsortedMatches", Params: []string{"container1", "container2"}},
+}
+
+func (c *deepUnsortedMatchChecker) Check(params []interface{}, _ []string) (bool, string) {
+	container1 := reflect.ValueOf(params[0])
+	container2 := reflect.ValueOf(params[1])
+
+	if container1 == container2 {
+		return true, ""
+	}
+
+	if container1.Kind() != container2.Kind() {
+		return false, fmt.Sprintf("containers are of different types: %s != %s", container1.Kind(), container2.Kind())
+	}
+
+	if container1.Len() != container2.Len() {
+		return false, fmt.Sprintf("containers %v and %v have different lengths: %d != %d",
+			container1.Interface(), container2.Interface(), container1.Len(), container2.Len())
+	}
+
+	switch container1.Kind() {
+	case reflect.Array, reflect.Slice:
+		return deepSequenceMatch(container1, container2)
+
+	case reflect.Map:
+		map1 := container1.Interface()
+		map2 := container2.Interface()
+		if !reflect.DeepEqual(map1, map2) {
+			return false, fmt.Sprintf("maps %v and %v don't match", map1, map2)
+		}
+
+		return true, ""
+
+	default:
+		return false, fmt.Sprintf("%T is not a supported container. Must be a slice, an array or a map", container1)
+
+	}
+}
+
+func deepSequenceMatch(container1 reflect.Value, container2 reflect.Value) (bool, string) {
+	matched := make([]bool, container1.Len())
+
+out:
+	for i := 0; i < container1.Len(); i++ {
+		el1 := container1.Index(i).Interface()
+
+		for e := 0; e < container2.Len(); e++ {
+			el2 := container2.Index(e).Interface()
+
+			if !matched[e] && reflect.DeepEqual(el1, el2) {
+				matched[e] = true
```

Comment:
can you add some comments to this code? how I think this code is working (please correct me if I'm wrong) is that we essentially take an element in container1, match it to something in container2 and mark the index we matched it with, so that we can't ever match multiple items in container1 with the same index in container2. But it took me a while to figure that out while staring at the code, so having a comment somewhere in the function would be good to folks like myself :smile:

Reply:
I'll add a comment but just wanted to say your description is exactly right. Marking each index as already matches prevents duplicate elements in container1 from being matched to the same element in container2 (which could then have elements not in container1 as long as the length was the same, e.g., [a, a, b] [a b c]).